flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java

```diff
@@ -145,8 +145,27 @@ protected Catalog createCatalog(
       baseNamespace = Namespace.of(properties.get(BASE_NAMESPACE).split("\\."));
     }
 
-    boolean cacheEnabled = Boolean.parseBoolean(properties.getOrDefault(CACHE_ENABLED, "true"));
-    return new FlinkCatalog(name, defaultDatabase, baseNamespace, catalogLoader, cacheEnabled);
+    boolean cacheEnabled = CatalogProperties.CACHE_ENABLED_DEFAULT;
+    if (properties.containsKey(CatalogProperties.CACHE_ENABLED)) {
+      cacheEnabled = Boolean.parseBoolean(properties.get(CatalogProperties.CACHE_ENABLED));
+    }
+
+    long cacheExpirationIntervalMs =
+        PropertyUtil.propertyAsLong(
+            properties,
+            CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS,
+            CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS_DEFAULT);
+    if (cacheExpirationIntervalMs == 0) {
```

Comment:
we should just do a `Preconditions` check here that this is a positive number.

Reply:
Not sure why limit to positive numbers, since negative values mean "cache expiration is turned off and entries expire only on refresh etc", users can set a negative values.

Reply:
> * If cache.expiration-interval-ms is zero, caching will be turned off entirely (irrespective of the cache-enabled flag)
> * If cache.expiration-interval-ms is positive (with cache-enabled=true), then caching will be on and entries will expire if not accessed in that period of time.
> * If users pass in a value <= -1, (with cache-enabled=true), then caching will be enabled and cache expiration will never happen (i.e. entries will persist unless explicitly refreshed, original behavior).

detail ref:  https://github.com/apache/iceberg/pull/3543#issue-1052281636

Reply:
@lvyanquan because we retained the boolean flag for enabling caching, I thought the refresh interval should just be positive. But apparently, I didn't get the implication of negative values. So forget about this comment.

Reply:
This PR replicates the SparkCatalog behavior, which is fine by me.

> - If cache.expiration-interval-ms is zero, caching will be turned off entirely (irrespective of the cache-enabled flag)
> - If users pass in a value <= -1, (with cache-enabled=true), then caching will be enabled and cache expiration will never happen (i.e. entries will persist unless explicitly refreshed, original behavior).

I find it a little weird that zero and negative numbers have different implications. why distinguish those two? @kbendick can you share some context?

Reply:
if we look at the code of `CachingCatalog`, it only check positive values. Hence don't understand why we need value zero to disable caching since we already have the boolean flag. `positive` refresh interval would be caching with expiration. otherwise, it is caching without expiration (until refreshed).

```
  private Cache<TableIdentifier, Table> createTableCache(Ticker ticker) {
    Caffeine<Object, Object> cacheBuilder = Caffeine.newBuilder().softValues();

    if (expirationIntervalMillis > 0) {
      return cacheBuilder
          .removalListener(new MetadataTableInvalidatingRemovalListener())
          .executor(Runnable::run) // Makes the callbacks to removal listener synchronous
          .expireAfterAccess(Duration.ofMillis(expirationIntervalMillis))
          .ticker(ticker)
          .build();
    }

    return cacheBuilder.build();
  }
```

Reply:
In my opinion, these two parameters actually have some overlap.
If we introduce a value of 0, we actually don't need ‘cache-enabled’, which is exactly what @hililiwei said before.
***
Can we drop cacheEnabled and just use cacheExpirationIntervalMs? When cacheExpirationIntervalMs is zero, it is disabled.https://github.com/apache/iceberg/pull/6111#discussion_r1012530287
***
Here is just for compatibility with the previous configuration.

Reply:
I understood the overlap btw the boolean flag and the expire value at 0. But dropping `cacheEnabled ` would be a breaking change, which we should try to avoid

Reply:
I think the key decision is whether to keep the same [parameter definitions](https://iceberg.apache.org/docs/latest/configuration/#catalog-properties) as before or to remove and verify unnecessary 0 values in Flink. I agree with the same configuration as Spark, because the parameter is public. 
But Flink has a separate [configuration page](https://iceberg.apache.org/docs/latest/flink/#catalog-configuration), so redefining is also acceptable.

Reply:
> But Flink has a separate [configuration page](https://iceberg.apache.org/docs/latest/flink/#catalog-configuration), so redefining is also acceptable.

I had the same thought. Here is what is intuitive to me.
* if the boolean flag is false, caching is disabled
* if the boolean flag is true
  * if  cacheExpirationIntervalMs > 0, caching automatically expired after the value
  * if <=0, caching never expire automatically

What are you takes? @lvyanquan @hililiwei @pvary

Reply:
Coming from a world where we were a platform for multiple companies, I would chose the more conservative approach and try to avoid breaking changes unless it is strictly necessary.

Also adding a new configuration which just provides another way to archive the same function seems something which will lead to confusion down the line, so I would opt for:
- keep the  `cache-enabled` boolean flag
- if the `cache-enabled` is `true` find meaningful values for the `cacheExpirationIntervalMs`, or throw an error message if they do not make sense, so:
   - if `cacheExpirationIntervalMs` is > 0 then expire the the element
   - if `cacheExpirationIntervalMs` is <=0 then never expire the element
   - if `cacheExpirationIntervalMs` is unset then never expire the element to keep the backwards compatibility
- if the `cache-enabled` is `false` then ignore the `cacheExpirationIntervalMs` or throw an exception if it is set.

Just my 2 cents

Reply:
Addressed for these suggestions:
>   * if `cacheExpirationIntervalMs` is unset then never expire the element to keep the backwards compatibility.    

1. Use CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS_OFF as default value to keep the same behavior.

>   * if <=0, caching never expire automatically. this is the difference. don'y disable caching with value 0 because boolean flag serves the purpose already.

2. Howerver, 0 value is excluded in the constructor of [CachingCatalog](https://github.com/apache/iceberg/blob/dbb8a404f6632a55acb36e949f0e7b84b643cede/core/src/main/java/org/apache/iceberg/CachingCatalog.java) as the following code：
```
  protected CachingCatalog(
      Catalog catalog, boolean caseSensitive, long expirationIntervalMillis, Ticker ticker) {
    Preconditions.checkArgument(
        expirationIntervalMillis != 0,
        "When %s is set to 0, the catalog cache should be disabled. This indicates a bug.",
        CatalogProperties.CACHE_EXPIRATION_INTERVAL_MS);
    this.catalog = catalog;
    this.caseSensitive = caseSensitive;
    this.expirationIntervalMillis = expirationIntervalMillis;
    this.tableCache = createTableCache(ticker);
  }
```
So cacheExpirationIntervalMs is not allowed to be 0, add this restriction to Precondition and Document too.

Reply:
latest code looks good to me. Will merge tomorrow unless there are other comments.

We are on the same page regarding backward compatibility for keeping the boolean flag.

> cacheExpirationIntervalMs is not allowed to be 0

agree with the Precondition.  on the other hand, I think it is is a bit unnecessary for the CachingCatalog. it could have treated `<=0` as no expiration.